Moving to interactive service

lstipakov opened the pull request.

lstipakov committed:
Split into multiple files

No functional changes.

GitHub: #19

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
GHA: enable for all branches

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
Decouple OpenVPN service logic from Windows service infrastructure

Move OpenVPN service logic into a separate
class which doesn't have any dependencies on
Windows service infrastructure.

Add a log delegate which is used to write
logs either to console or event log.

Add ability to run service logic
as a standalone app.

GitHub: #21

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
Use interactice service to start openvpn process

Instead of starting openvpn process directly:

 - connect to interactive service pipe
 - send startup info
 - read openvpn process pid (for polling)

Since we use virtual service account, we don't have
privileges to subscribe for process events, so instead
we use async polling to check if process is alive and
restart it of it is not.

Removed Stop() since due to lack of privileges
the only way now we can stop the process is to
signal the event.

GitHub: #23

Signed-off-by: Lev Stipakov <lev@openvpn.net>

selvanair reviewed (commented):
Its a bit hard to review this with all the refactoring going on. Would it be possible to split this with the refactoring and cleanup moved to a future PR? I thought the last commit would pretty much covers all that is needed for the iservice-based startup to work.

lstipakov commented:
> Its a bit hard to review this with all the refactoring going on. Would it be possible to split this with the refactoring and cleanup moved to a future PR? I thought the last commit would pretty much covers all that is needed for the iservice-based startup to work.

But my idea was exactly this - there are few commits for refactoring without any functional changes (like splitting functionality into different files) and this last commit is switching to iservice.

lstipakov commented on OpenVPNChild.cs:
```diff
@@ -193,21 +158,46 @@ private void FastRestart(object sender, EventArgs e)
+                    // send startup info
+                    var logOption = config.logAppend ? "--log-append " : "--log";
+                    var cmdLine = $"--config \"{configFile}\" {logOption} \"{logFile}\" --service \"{exitEvent}\" 0 --pull-filter ignore route-method";
+
```
Done!

lstipakov replied:
Apparently not. Just tested this:

`--config c:\\Users\\lev\\OpenVPN\\config\\server\\server.ovpn --log c:\\Temp\\log1.txt --log c:\\Temp\\log2.txt `

and the `log1.txt` was created.

lstipakov committed:
Remove unused functions

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
Remove priority class code

Priority class cannot be set by the
virtual service account so remove the code.

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
Remove installer code

The service is installed via MSI Wix script.

GitHub: #24

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
Generate assembly metadata during build

Instead of having version number in multiple places,
keep it in one place and generate AssemblyInfo.cs on the fly.

While on it, add git commit hash to product version
and bump version to 2.0.0.0.

GitHub: #25

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov committed:
mono: switch to MSBuild

Mono version coming with Ubuntu 22.04 is an outdated
and have bugs like inability to handle escaped
semicolons (https://github.com/mono/mono/pull/1580/files).

Bump to Mono 6.12 and switch to MSBuild (since xbuild is outdated).

Signed-off-by: Lev Stipakov <lev@openvpn.net>

lstipakov commented:
[Here](https://github.com/lstipakov/openvpn-build/actions/runs/13835718677) you can get an installer with necessary changes to all components. At least here it works fine :)

d12fk commented:
I think this service should start the interactive service, if it is not running already.

lstipakov commented:
> I think this service should start the interactive service, if it is not running already.

It is now depended on interactive service https://github.com/OpenVPN/openvpn-build/pull/849/files so Windows should start iservice first.

d12fk commented on .github/workflows/build.yml:
```diff
@@ -2,11 +2,7 @@ name: OpenVPNServ2
 on:
   push:
-    branches:
-     - master
```
Is this what we want in master? Is this code not built in gerrit?

lstipakov replied:
I think only openvpn project itself is built in gerrit and the rest (including this) are not. I think GHA should be called for all branches, not only master.

d12fk commented on Program.cs:
```diff
@@ -0,0 +1,53 @@
+    {
+        public static int Main(string[] args)
+        {
+            if (args.Length == 0)
```
The command line args could be more like with the interactive service to make life easier for the uninitiated. Especially the lack of a command line help was noticed and deemed missing by me.

lstipakov replied:
I think we already covered it - when you run it as is, it just works as a standalone app which talks to interactive service (like GUI) - this part comes in follow-up commit. If you want to install/run service, then sc is a way to go.

d12fk commented on OpenVPNServiceRunner.cs:
```diff
@@ -0,0 +1,181 @@
+
+namespace OpenVpn
+{
+    class OpenVPNServiceRunner
```
Are there functional changes hidden in here besides the stuff related to the new Log() method?

d12fk reviewed (commented):
Think this could generally benefit from connecting to the management interface, just like the GUI does. Instead of polling, I mean.

d12fk commented on OpenVPNChild.cs:
```diff
@@ -27,43 +29,14 @@ public OpenVpnChild(OpenVpnServiceConfiguration config, string configFile)
+            var justFilename = Path.GetFileName(configFile);
+            logFile = Path.Combine(config.logDir, justFilename.Substring(0, justFilename.Length - config.configExt.Length) + ".log");
 
             // FIXME: if (!init_security_attributes_allow_all (&sa))
```
What's this?

lstipakov replied:
It appears to be a leftover from a previous C-based implementation (?). This commit just splits code into multiple files without any functional changes.

d12fk replied:
Why not remove it if it's not needed anymore, this is the time.

lstipakov replied:
Well, hard to disagree. Removed!

d12fk commented on OpenVPNServiceConfiguration.cs:
This is mentioned in the previous commit message, but ultimately done here. Please update the commit message there.

lstipakov replied:
Hm, I don't get it. What is mentioned in the previous (in what exactly) commit message?

This Log delegate is needed to support logging to console when running as a standalone app and to the event log when running as a service.

d12fk replied:
Hm, somehow I added this comment in the wrong place. It's basically about removing the sentence "Removed setting priority class for the same reason." from the commit message for 7fd4ceb7, because that's done in b2fe5163

lstipakov replied:
Got it, fixed.

d12fk commented on Program.cs:
```diff
@@ -11,57 +11,22 @@ internal class Program
-                    runner.Stop();
-                }
-            }
-            else if (args[0] == "-install")
```
Having the install is useful when you want to run the service from the build dir without copying all the time. That's my workflow for the interactive service.

lstipakov replied:
You are the second person mentioning this, so it makes me think twice. 

You (=in general) could do install/remove/start/stop with `sc.exe` or powershell. You might need to type a bit more, but at least with powershell there is an autocompletion. You could also create a simple .bat file for that. Besides, this could be now used ad a standalone app so no need to install/remove the service (at least if you test certain parts of code). 

Does this small convenience really worth keeping 300+ lines of code, which is not really used in production?

 ```
OpenVpnService.csproj        |   9 ---
 Program.cs                   |  59 ++++----------------
 ProjectInstaller.Designer.cs |  61 --------------------
 ProjectInstaller.cs          | 100 ---------------------------------
 ProjectInstaller.resx        | 129 -------------------------------------------
 5 files changed, 12 insertions(+), 346 deletions(-)
```

d12fk replied:
Is it 300 lines of code to register with the SCM? In that case a script to install it using sc or ps is probably better.

lstipakov replied:
Yeah there is a separate commit ("Remove installer code") which removes a lot of strange code. This service was written in pre-MSI age, although it still doesn't explain why one couldn't use sc.exe in NSIS script.

lstipakov replied:
> Yeah there is a separate commit ("Remove installer code") which removes a lot of strange code. This service was written in pre-MSI age, although it still doesn't explain why one couldn't use sc.exe in NSIS script.

Oh well, this separate commit is what we are looking now. But as you can see it has removed a lot of installer code, which is now handled by MSI installer.

NiKiZe replied:
In framework it is a lot of code, in core, it is like 10 lines.

d12fk commented:
Doxygen would be nice, but a bit more excessive commit messages, describing the things done in the commit, are missing often.

d12fk commented on build.sh:
mono in Ubuntu comes with xbuild and not msbuild. I would suspect the same for other distributions.

lstipakov commented on OpenVPNChild.cs:
```diff
@@ -108,66 +82,93 @@ public void SignalProcess()
-        // terminate process after a timeout
-        public void StopProcess(int timeout)
+        /// **Polling loop to monitor process exit**
+        private async void MonitorProcessExit()
```
What do you mean exactly by monitoring? We could probably poll it but then what? If interactive service crashes we don't have permissions to restart it (like we do for openvpn.exe) since we use very limited user account. I think it would be better to change the installer and configure Windows to auto-restart the failing service.

d12fk replied:
Yeah, that is indeed a better solution then.
